src/mpid/ch4/src/ch4_coll.h

```diff
@@ -409,9 +409,12 @@ MPL_STATIC_INLINE_PREFIX int MPID_Allreduce(const void *sendbuf, void *recvbuf,
     is_commutative = MPIR_Op_is_commutative(op);
 
     /* FIXME: this fallback check assumes device algorithms are limited by release gather shm cell size */
-    MPI_Aint dt_size;
+    MPI_Aint dt_size, lb, extent, true_extent;
     MPIR_Datatype_get_size_macro(datatype, dt_size);
-    if (dt_size >=
+    MPIR_Type_get_extent_impl(datatype, &lb, &extent);
+    MPIR_Type_get_true_extent_impl(datatype, &lb, &true_extent);
+    extent = MPL_MAX(extent, true_extent);
+    if (MPL_MAX(dt_size, extent) >=
         MPIR_CVAR_REDUCE_INTRANODE_BUFFER_TOTAL_SIZE / MPIR_CVAR_REDUCE_INTRANODE_NUM_CELLS) {
         goto fallback;
     }
```

Comment:
Come to think of it, this check is still not sufficient. Since datatype information is local, the extent may be uniform across processes. We really need to do the reduction in terms of basic type, otherwise we risk deadlock or data corruption.

Reply:
I am not sure I follow. All processes have to use the same datatype, right?

Reply:
I initially thought that different layouts might be allowed, but looking closely at the standard it seems to be more restricted.

> REDUCE: The routine is called by all group members using the same arguments for
count, datatype, op, root and comm. Thus, all processes provide input buffers of the same
length, with elements of the same type as the output buffer at the root.

ALLREDUCE has this caveat for intercommunicators that seems to indicate that data layouts can only differ between groups and not within the group:

> If comm is an inter-communicator, then the result of the reduction of the data provided
by processes in group A is stored at each process in group B, and vice versa. Both groups
should provide count and datatype arguments that specify the same _type signature_.

Reply:
I don't think `realease_gather` works with inter-communicator